### DISCUSSION TITLE
Trying it out again, got some questions!

### DISCUSSION THREAD

**Discussion by fideloper** [php/frankenphp#346] [Q&A] [answered]
Hey there!

Frankenphp has improved so much, this looks great!

I have a few questions, I wasn't sure about them when looking at the docs!

1. Laravel seems to work (locally for me) by default - worker mode (and Octane) seems to be optional - is that correct? Or does this require Octane for Laravel to work correctly? (I saw the PR which isn't yet merged, I wasn't sure if Laravel was just working by coincidence since there's a single user - me - running it locally).
2. Is Caddy required, right? (vs Nginx proxying to the static binary?) I like Caddy a lot but am curious about if its needed!
3. Does the Go side of things give us any hooks to use (I think RoadRunner does, for example, to help support grpc)

Really digging it!

**Comment by EvanEdwards** [comment]
2.  Caddy is not required.  FrankenPHP is at heart a Caddy binary with several large extensions compiled in.

A FrankenPHP binary will, however, fit in nicely with a preexisting systemd, caddy user and group, and /etc/caddy/Caddyfile as a completely compatible drop in _**replacement**_ for caddy on many systems with just a couple lines edited (or, as I did, copying caddy.service to frankenphp.service and editing it to point to the new binary).

**Reply by fideloper** [reply]
Thanks! Just to be totally clear, the above says `Caddy is **not** required.` but the rest of it sounds like Caddy **is** required (which I'm going to assume is the case) 😂

**Reply by EvanEdwards** [reply]
From a _practical standpoint as of today,_ FrankenPHP consists of {Caddy, PHP, Mercure, Prometheus, Vulcain, worker mode, access to Early Hints} all compiled into one standalone binary.  Possibly some more features, as that's off the top of my head.  

FrankenPHP _the project_ is as @dunglas says: a PHP app server that works with Go.  That likely means there will be other options in the future.

To put it another way (I'm using a path just to indicate when I'm talking about actual compiled binaries): you do not need /bin/caddy, as /bin/frankenphp fully replaces it with a superset of features.  To compile /bin/frankenphp from source, you _will_ need the Caddy sourcecode.  However, if you have a server or development laptop, you do not need /bin/caddy at all.  /bin/frankenphp includes webserver+PHP and more, all in one binary.

Static binaries of /bin/frankenphp are available, so you can drop it on a fresh system and it will launch a webserver with PHP support.  There is no need to have ever installed Caddy on that system to have it be fully functional.

**Comment by dunglas** [comment, marked as answer]
Hi @fideloper,

Thanks for the kind words, it means a lot!

> Laravel seems to work (locally for me) by default - worker mode (and Octane) seems to be optional - is that correct? Or does this require Octane for Laravel to work correctly? (I saw the PR which isn't yet merged, I wasn't sure if Laravel was just working by coincidence since there's a single user - me - running it locally).

Laravel is indeed entirely supported. Worker mode/Octane is entirely optional.
(Octane integration is almost done, but not merged yet: https://github.com/laravel/octane/pull/764)

> Is Caddy required, right? (vs Nginx proxying to the static binary?) I like Caddy a lot but am curious about if its needed!

FrankenPHP is made of several components:

* At its heart, it's a standalone Go library that does not require Caddy, and can be used with the built-in Go server. Here is an example of how to create a server without Caddy: https://pkg.go.dev/github.com/dunglas/frankenphp#example-ServeHTTP
* A Caddy module is also provided. In the future, we may develop modules for other servers written in Go such as Traeffik.
* Finally, we provide Docker images and static binaries that are basically Caddy compiled with the FrankenPHP Caddy module (which uses the FrankenPHP Go library) + PHP (libphp provided by "official" PHP Docker images with Docker, and a static version of libphp created with `static-php-cli` for the static binaries).

TL;DR: we choose to ship Caddy by default instead of just the built-in Go web server because it adds a lot of cool features, but it's totally possible to build a "minimal" server not using Caddy. We even have one in our test suite: https://github.com/dunglas/frankenphp/blob/main/internal/testserver/main.go

> Does the Go side of things give us any hooks to use (I think RoadRunner does, for example, to help support grpc)

Do you have a link to the hooks provided by RoadRunner? What do you mean by "GRPC support", I'm not very familiar with grpc in PHP, but isn't the official grpc extension enough? We provide some extension points that should be enough for some use cases (https://pkg.go.dev/github.com/dunglas/frankenphp#pkg-overview) but we would be happy to add more if needed!

**Reply by fideloper** [reply]
Thanks so much, that clarifies a bunch!

The gRPC point is less important - php falls down a bit on grpc (it can be a client but not a server), I think roadrunner has fancy way to make up for that (not 100% sure!). No big deal there (I barely have time to look into it and am def not asking you to :D )

**Comment by nguyenthehiepsi** [comment]
@dunglas 
I am facing problem with franken worker mode for laravel integration. I am very appricated with your help. The problem is as below:

I am building a very big system which i need nginx as a front, i have made config as laravel document
https://laravel.com/docs/11.x/octane

Here is franken json for worker mode
```
{
	# Enable FrankenPHP
	frankenphp {
		worker ./public/index.php 4
	}
}
```

Here is franken.services defination on systemd
```
[Unit]
Description=FrankenPHP Service
After=network.target

[Service]
WorkingDirectory=/var/www/web
Environment="HOME=/root"
Environment="XDG_CONFIG_HOME=$HOME/.config"
Environment="FRANKENPHP_CONFIG="worker ./public/index.php"
ExecStart=/usr/bin/frankenphp run -c /var/www/web/frankenphp.json -a caddyfile
#Restart=always
Restart=on-failure
PIDFile=/run/frankenphp/frankenphp.pid
StandardOutput=append:/var/log/frankenphp/frankenphp.log
StandardError=append:/var/log/frankenphp/frankenphp-err.log

[Install]
WantedBy=multi-user.target
```

franken works very well without worker mode but when worker mode is enabled, it could not resolve laravel redirect correctly (laravel use redirect() method). 
I have researched a lot and have some glue but i can not find a solution.
With laravel Auth for login/logout, Auth is using session driver redis as well.
`
After logout from web service, i had a flow log as example below:
/logout (Auth:logout) → Redirect to /login (Auth:check) → Redirect back to /toppage (login back automatically after logout because Auth:check passed even though logout has just requested) 

As i understand, laravel redirect route is not resolve correctly (nginx or franken does has misconfigure) or Auth is not processed correctly on franken.

**Reply by fideloper** [reply]
re: Redirects: Is APP_URL set correctly? I wonder if that's needed in worker (octane) mode.

You didn't say HOW the redirects were failing - is it the domain, or http vs https or something else?

If it's http/https, you may need to set trusted proxies to trust `*` (everything).

**Reply by nguyenthehiepsi** [reply]
@fideloper 

> is it the domain, or http vs https or something else?
>If it's http/https, you may need to set trusted proxies to trust * (everything).

APP_URL=https://xxxx ←domain site is defined here, its not "localhost" or "127.0.0.1"

Do I have to set it to "locahost" or "127.0.0.1"
APP_URL=http://localhost:9000, which is correct settings? Can you help to clarify?

> You didn't say HOW the redirects were failing
trusted proxies is set by default by laravel and it was for all (*)


After logout from web service, i had a flow log as example below:
/logout (Auth:logout) → Redirect to /login (Auth:check) → Redirect back to /toppage (login back automatically after logout because Auth:check passed even though logout has just requested)


>  I wonder if that's needed in worker (octane) mode.

I have changed/replaced  [./public/index.php] for working with frankenphp worker mode but actually i dont know which index.php content is correct or not
As i see, frankenphp worker mode is run but not compatible to all the laravel routes or functtions as i already posted above

i  found lots of isssues related frankenphp with nginx upstream even though franken is not set in worker mode
https://github.com/dunglas/frankenphp/issues/557

I also meet the same problems but i dont know how to solve 
here is my nginx config for proxy pass
```
upstream frankenphp_backend{
	server 127.0.0.1:9000;
}

server {
location /index.php {
		try_files /not_exists @octane;
	}

	location / {
		try_files $uri $uri/ @octane;
	}


	location @octane {
		set $suffix "";

		if ($uri = /index.php) {
			set $suffix ?$query_string;
		}

		proxy_buffering on;
		proxy_http_version 1.1;
		proxy_set_header Host $http_host;
		proxy_set_header Scheme $scheme;
		
		proxy_set_header SERVER_PORT $server_port;
		proxy_set_header REMOTE_ADDR $remote_addr;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		#proxy_set_header X-Forwarded-Proto $scheme;
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header Upgrade $http_upgrade;
		proxy_set_header Connection $connection_upgrade;
		
		#frankenphp proxy timeout is nessary
		proxy_read_timeout 300s;   # Adjust the read timeout
                proxy_connect_timeout 300s; # Adjust the connect timeout
               proxy_send_timeout 300s;    # Adjust the send timeout

		proxy_buffer_size 128k;
		proxy_buffers 64 128k;			
		proxy_busy_buffers_size 256k;
		proxy_max_temp_file_size 512m;
                 proxy_temp_file_write_size 128k;	
		
		#run with upstream
		proxy_pass http://frankenphp_backend$suffix;
	}
}
```

**Comment by nguyenthehiepsi** [comment]
@dunglas 
Processes are tracking as attached images
<img width="1185" alt="スクリーンショット 2024-08-16 10 45 00" src="https://github.com/user-attachments/assets/149dfe58-da70-4ae1-9c80-cddb65aed93e">

Can you help to explain 2 of these?What is "caddy run"? is it created as a must?

**Comment by nguyenthehiepsi** [comment]
@dunglas 

frankenphp is set to run with root user, by the way, it can not run "get_current_user() ", is there any related issue which laravel app is loaded on memory and when run by franken, some method can not be exucuted?

Below is laravel's "config/logging.php"
```
'daily' => [
            'driver' => 'daily',
            'path' => storage_path('logs/' . get_current_user() . '-laravel.log'),
            'level' => env('LOG_LEVEL', 'debug'),
            'days' => 7, // SUPERCOLONY_797 - 20211116
            'permission' => 0777,
        ],
```

**Comment by nguyenthehiepsi** [comment]
@fideloper cc @dunglas 
For futher information, my php file call to "php artísan" command and after a short of time i have below dump error
when i  check by "journalctl -p 3", there are lots of systemd-coredumps

```
Process 400222 (frankenphp) of user 0 dumped core.
                                                                                          
            Stack trace of thread 400232:
            #0  0x0000ffff8162f3a8 n/a (/usr/bin/frankenphp + 0x82f3a8)
            #1  0x0000ffff8160bf84 n/a (/usr/bin/frankenphp + 0x80bf84)
            #2  0x0000ffff8160c39c n/a (/usr/bin/frankenphp + 0x80c39c)
            #3  0x0000ffff8160acd4 n/a (/usr/bin/frankenphp + 0x80acd4)
            #4  0x0000ffff8162f774 n/a (/usr/bin/frankenphp + 0x82f774)
            #5  0x0000ffff8402124c n/a (/usr/bin/frankenphp + 0x322124c)
            #6  0x0000ffff83ffd36c n/a (/usr/bin/frankenphp + 0x31fd36c)
            #7  0x0000ffff82fd0944 n/a (/usr/bin/frankenphp + 0x21d0944)
            #8  0x0000ffff2f4f1a40 n/a (n/a + 0x0)
            #9  0x0000ffff2f4f1a40 n/a (n/a + 0x0)
            #10 0x4ada39cc00000004 n/a (n/a + 0x0)
            ELF object binary architecture: AARCH64
```

my server run on t4g.small (amazon linux 20223 arm64) which has 2G memory, is there any problem with this low specs?
I wonder how many resources are needed for franken?

I am very appreciated with your answer, hope to have your answers soon

**Comment by nguyenthehiepsi** [comment]
@dunglas @fideloper 
Updated!

I tried lots of different server's specs. With same php process, franken used of one of CPU cores and gave the "systemd-coredump[2647]: Process 2416 (frankenphp) of user 0 dumped core." error as my last comment above
![スクリーンショット 2024-08-21 19 05 51](https://github.com/user-attachments/assets/a7fed912-a447-48b5-9418-ba64c5efade6)

this problem did not happen with php-fpm

As above comment, i am comparing btw nginx+php-fpm and nginx+frankenphp

I hope we can have some deeper discussion.
Thanks

**Reply by dunglas** [reply]
You may be hitting a known issue with OpenSSL and static builds: https://frankenphp.dev/docs/known-issues/#buggy-php-extensions

You can confirm this by gathering a stack trace with a debug build: https://frankenphp.dev/docs/contributing/#debugging-segmentation-faults-with-static-builds

If it's the case, switching to Docker images or compiling FrankenPHP locally will fix the issue. The PHP team is working on fixing this one, if you gather a stack trace, please post it on the PHP bug report, it may help.

**Reply by nguyenthehiepsi** [reply]
@dunglas 
thanks indeed for your repply

>You may be hitting a known issue with OpenSSL and static builds: https://frankenphp.dev/docs/known-issues/#buggy-php-extensions

I don't think it related to OPENSSL, my service is built on aws and backend is constructed with nginx+franken, also i don't find any error msg written on sharing link.

**Reply by dunglas** [reply]
Can you please gather a stack trace with a debug build (link in the previous comment), so we'll have a better idea of what's going on?

**Reply by nguyenthehiepsi** [reply]
@dunglas 
plz wait a moment, i don't use docker on production, so i am rebuilding binary static builder today, having problem with installation with libraries "musl-wrapper musl-cross-make" on aarch64 base
I am trying to contact you asap

**Reply by dunglas** [reply]
You shouldn't have to create your own build. Downloading the `-debug` variant from here https://github.com/dunglas/frankenphp/releases/tag/v1.2.4 should be good enough.

**Reply by nguyenthehiepsi** [reply]
@dunglas 
I do not know exactly this is what you want to look into.

franken version:
```
$ frankenphp -v
FrankenPHP v1.2.4 PHP 8.3.10 Caddy v2.8.4 h1:q3pe0wpBj1OcHFZ3n/1nl4V4bxBrYoSoab7rL9BMYNk=
```

and also frankenphp.json without worker mode:
```
{
	# Global options block
	debug # Enable debug mode for more verbose logs
	log {
		output file /var/log/frankenphp/frankenphp-err.log {
			roll_size 5mb
			roll_keep 3
			roll_keep_for 720h
		}
		level ERROR # Set the log level (DEBUG, INFO, WARN, ERROR)
	}

	# Enable FrankenPHP
	frankenphp {
		#{$FRANKENPHP_CONFIG}
	}
}

:9000 {
	root * projectxxx/public

	# Execute PHP files in the current directory and serve assets
	php_server {
		# Required for the public/storage/ dir
		resolve_root_symlink
	}
	encode zstd br gzip
	file_server
}
```

maybe gdb command result is not enough


```
$ coredumpctl gdb 424013
           PID: 424013 (frankenphp)
           UID: 0 (root)
           GID: 0 (root)
        Signal: 11 (SEGV)
     Timestamp: Thu 2024-08-22 12:24:14 UTC (12min ago)
  Command Line: /usr/bin/frankenphp run -c ./frankenphp.json -a caddyfile
    Executable: /usr/bin/frankenphp
 Control Group: /system.slice/frankenphp.service
          Unit: frankenphp.service
         Slice: system.slice
       Boot ID: 2176899d50784cac853e4b72702d47cb
    Machine ID: ec2c63b370247b9b530d00fc067b180c
      Hostname: ip-xxxx.ap-northeast-1.compute.internal
       Storage: /var/lib/systemd/coredump/core.frankenphp.0.2176899d50784cac853e4b72702d47cb.424013.1724329454000000.xz (present)
  Size on Disk: 12.5M
       Message: Process 424013 (frankenphp) of user 0 dumped core.             
  
              Stack trace of thread 424025:
                #0  0x0000ffffa60303a8 n/a (/usr/bin/frankenphp + 0x8303a8)
                #1  0x0000ffffa600cf84 n/a (/usr/bin/frankenphp + 0x80cf84)
                #2  0x0000ffffa600d39c n/a (/usr/bin/frankenphp + 0x80d39c)
                #3  0x0000ffffa600bcd4 n/a (/usr/bin/frankenphp + 0x80bcd4)
                #4  0x0000ffffa6030774 n/a (/usr/bin/frankenphp + 0x830774)
                #5  0x0000ffffa8a267cc n/a (/usr/bin/frankenphp + 0x32267cc)
                #6  0x0000ffffa8a028ec n/a (/usr/bin/frankenphp + 0x32028ec)
                #7  0x0000ffffa79d1224 n/a (/usr/bin/frankenphp + 0x21d1224)
                #8  0x0000ffff5353e250 n/a (n/a + 0x0)
                #9  0x0000ffff5353e250 n/a (n/a + 0x0)
                ELF object binary architecture: AARCH64

GNU gdb (GDB) Amazon Linux 12.1-5.amzn2023.0.3
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-amazon-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/frankenphp...
(No debugging symbols found in /usr/bin/frankenphp)
warning: Can't open file /dev/zero (deleted) during file-backed mapping note processing
[New LWP 424025]
[New LWP 424024]
[New LWP 424017]
[New LWP 424018]
[New LWP 424020]
[New LWP 424021]
[New LWP 424022]
[New LWP 424023]
[New LWP 424495]
[New LWP 424019]
[New LWP 424013]
Core was generated by `/usr/bin/frankenphp run -c ./frankenphp.json -a caddyfile'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000ffffa60303a8 in ?? ()
[Current thread is 1 (LWP 424025)]
(gdb) bt
#0  0x0000ffffa60303a8 in ?? ()
#1  0x0000ffffa600d38c in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb) 
```

I am so sorry if i do not use gdb correctly as log has "(No debugging symbols found in /usr/bin/frankenphp)"
I will update more after having your instructions.

**Reply by dunglas** [reply]
It looks like you aren't running the debug binary (the file suffixed by `-debug` in the assets section).

**Reply by nguyenthehiepsi** [reply]
@dunglas 
OS Version, is it possible to use "[frankenphp-linux-x86_64-debug] (https://github.com/dunglas/frankenphp/releases/download/v1.2.4/frankenphp-linux-x86_64-debug)" cose my sys is aarch64? 
```
$ uname -m
aarch64

$ cat /etc/os-release 
NAME="Amazon Linux"
VERSION="2023"
ID="amzn"
ID_LIKE="fedora"
VERSION_ID="2023"
PLATFORM_ID="platform:al2023"
PRETTY_NAME="Amazon Linux 2023.4.20240513"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2023"
HOME_URL="https://aws.amazon.com/linux/amazon-linux-2023/"
DOCUMENTATION_URL="https://docs.aws.amazon.com/linux/"
SUPPORT_URL="https://aws.amazon.com/premiumsupport/"
BUG_REPORT_URL="https://github.com/amazonlinux/amazon-linux-2023"
VENDOR_NAME="AWS"
VENDOR_URL="https://aws.amazon.com/"
SUPPORT_END="2028-03-15"
```

**Reply by dunglas** [reply]
Indeed, we don't provide debug builds for ARM (we'll do it when ARM runner for GitHub Actions is freely available, at the end of the year).

You'll have to compile a build with debug symbols for ARM yourself :/ The easier way is probably to compile it directly on the ARM server using the builder Docker image we provide: https://frankenphp.dev/docs/static/#linux

Please set `DEBUG_SYMBOLS` to `1` to keep debug symbols.

**Reply by nguyenthehiepsi** [reply]
@dunglas 
i have been struggling alot this afternoon as i told you on my previous comment, it doesnt work as well
I could not build for my own use, it may takes me more time to dig into the problem, below is where i was stopped at

```
make: * [Makefile:614: ext/fileinfo/libmagic/apprentice.lo] エラー 1
make: * 未完了のジョブを待っています....
[05:33:27] [ERRO] Uncaught SPC\exception\RuntimeException: Command run failed with code[2]: cd '/home/ec2-user/frankenphp/dist/static-php-cli/source/php-src' && make -j2 INSTALL_ROOT=/home/ec2-user/frankenphp/dist/static-php-cli/buildroot EXTRA_CFLAGS='-g -Os -fno-ident -fPIE' EXTRA_LIBS='/home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/liblz4.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libzip.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libfreetype.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libbz2.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libjpeg.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libturbojpeg.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libwebp.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libwebpdecoder.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libwebpdemux.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libwebpmux.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libsharpyuv.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libavif.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libpng16.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libcurl.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libnghttp2.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libiconv.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libcharset.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libbrotlidec.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libbrotlienc.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libbrotlicommon.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libssl.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libcrypto.a /home/ec2-user/frankenphp/dist/static-php-cli/buildroot/lib/libz.a  ' EXTRA_LDFLAGS_PROGRAM='-all-static' install at /home/ec2-user/frankenphp/dist/static-php-cli/src/globals/functions.php(147)
[05:33:27] [ERRO] #0 /home/ec2-user/frankenphp/dist/static-php-cli/src/SPC/util/UnixShell.php(51): f_passthru()
#1 /home/ec2-user/frankenphp/dist/static-php-cli/src/SPC/builder/linux/LinuxBuilder.php(300): SPC\util\UnixShell->exec()
#2 /home/ec2-user/frankenphp/dist/static-php-cli/src/SPC/builder/linux/LinuxBuilder.php(196): SPC\builder\linux\LinuxBuilder->buildEmbed()
#3 /home/ec2-user/frankenphp/dist/static-php-cli/src/SPC/command/BuildCliCommand.php(173): SPC\builder\linux\LinuxBuilder->buildPHP()
#4 /home/ec2-user/frankenphp/dist/static-php-cli/src/SPC/command/BaseCommand.php(101): SPC\command\BuildCliCommand->handle()
#5 /home/ec2-user/frankenphp/dist/static-php-cli/vendor/symfony/console/Command/Command.php(326): SPC\command\BaseCommand->execute()
#6 /home/ec2-user/frankenphp/dist/static-php-cli/vendor/symfony/console/Application.php(1078): Symfony\Component\Console\Command\Command->run()
#7 /home/ec2-user/frankenphp/dist/static-php-cli/vendor/symfony/console/Application.php(324): Symfony\Component\Console\Application->doRunCommand()
#8 /home/ec2-user/frankenphp/dist/static-php-cli/vendor/symfony/console/Application.php(175): Symfony\Component\Console\Application->doRun()
#9 /home/ec2-user/frankenphp/dist/static-php-cli/bin/spc(21): Symfony\Component\Console\Application->run()
#10 {main}
```

As i am experienced, these are lots of libraries/dependencies need to be installed before make build
```
bison flex patch cmake musl-wrapper musl-cross-make
```

i could not install "musl-wrapper musl-cross-make" so i could not move over for static build in order to make a binary debug verrsion

**Reply by dunglas** [reply]
Did you use the Docker image we provide? Everything, including the musl wrapper should be included in it.

**Reply by nguyenthehiepsi** [reply]
@dunglas 
Sorrry for reply late yesterday

I am using below command but it seems have errror. I do not know how to build "static-builder' locally, i may misreading from your document, can you help to check it out? Also, which is suitable content for "docker-bake.hcl"

```
$ sudo docker buildx bake --load --set static-builder.args.DEBUG_SYMBOLS=1 --set "static-builder.platform=linux/amd64" static-builder 
[+] Building 0.0s (1/1) FINISHED                                                                                                                                                                                                                                                                                                                             docker:default
 => [internal] load local bake definitions                                                                                                                                                                                                                                                                                                                             0.0s
 => => reading docker-bake.hcl 121B / 121B                                                                                                                                                                                                                                                                                                                             0.0s
ERROR: could not find any target matching 'static-builder'
```

if i run this alone, it worked well, but without the first one, it was not built in debug mode
```
$ docker cp $(docker create --name static-builder dunglas/frankenphp:static-builder):/go/src/app/dist/frankenphp-linux-$(uname -m) frankenphp
 Successfully copied 131MB to /home/ec2-user/frankenphp
```

here is my docker-bake.hcl defination
```
$ cat docker-bake.hcl
group "default" {
  targets = ["app"]
}

target "app" {
  dockerfile = "Dockerfile"
  tags = ["static-builder:latest"]
}

```
